iroh/src/baomap/flat.rs

```diff
@@ -0,0 +1,1210 @@
+                        .open(path.clone())
+                })
+                .await?;
+                writer.write_at(0, &size.to_le_bytes()).await?;
+                Ok(PreOrderOutboard {
+                    root: hash,
+                    tree,
+                    data: writer,
+                })
+            } else {
+                Err(io::Error::new(
+                    io::ErrorKind::PermissionDenied,
+                    "cannot write to in-memory outboard",
+                ))
+            }
+        }
+        .boxed()
+    }
+
+    fn data_writer(&self) -> BoxFuture<'_, io::Result<<Store as PartialMap>::DataWriter>> {
+        let data = self.entry.data.clone();
+        async move {
+            if let Either::Right((path, _)) = data {
+                let writer = iroh_io::File::create(move || {
+                    std::fs::OpenOptions::new()
+                        .write(true)
+                        .create(true)
+                        .open(path.clone())
+                })
+                .await?;
+                Ok(writer)
+            } else {
+                Err(io::Error::new(
+                    io::ErrorKind::PermissionDenied,
+                    "cannot write to in-memory data",
+                ))
+            }
+        }
+        .boxed()
+    }
+}
+
+impl PartialMap for Store {
+    type OutboardMut = PreOrderOutboard<File>;
+
+    type DataWriter = iroh_io::File;
+
+    type PartialEntry = Entry;
+
+    fn get_partial(&self, hash: &Hash) -> Option<Self::PartialEntry> {
+        let entry = self.0.state.read().unwrap().partial.get(hash)?.clone();
+        Some(Entry {
+            hash: blake3::Hash::from(*hash),
+            entry: EntryData {
+                data: Either::Right((
+                    self.0.options.partial_data_path(*hash, &entry.uuid),
+                    entry.size,
+                )),
+                outboard: Either::Right(self.0.options.partial_outboard_path(*hash, &entry.uuid)),
+            },
+        })
+    }
+
+    fn get_or_create_partial(&self, hash: Hash, size: u64) -> io::Result<Entry> {
+        let mut state = self.0.state.write().unwrap();
+        let entry = state.partial.entry(hash).or_insert_with(|| {
+            let uuid = rand::thread_rng().gen::<[u8; 16]>();
+            PartialEntry::new(size, uuid)
+        });
+        let data_path = self.0.options.partial_data_path(hash, &entry.uuid);
+        let outboard_path = self.0.options.partial_outboard_path(hash, &entry.uuid);
+        Ok(Entry {
+            hash: blake3::Hash::from(hash),
+            entry: EntryData {
+                data: Either::Right((data_path, size)),
+                outboard: Either::Right(outboard_path),
+            },
+        })
+    }
+
+    fn insert_complete(&self, entry: Entry) -> BoxFuture<'_, io::Result<()>> {
+        let hash = entry.hash.into();
+        let Either::Right((temp_data_path, size)) = entry.entry.data else {
+            todo!()
```

Comment:
what about these `todo`s?

Reply:
this is basically `unreachable!`. I just forgot to change it. I am using the same Entry type for partial and non-partial, but a partial entry can never have the left side. Two ways to solve this - either change it so there is a separate PartialEntry type, or return an io error "impossible" here.

Reply:
I mean `unreachable!` with a comment why it won't happen works well enough I would say

Reply:
OK, will do `unreachable!` now and then maybe have a separate entry type later.